### PR TITLE
`json`: document schema conversion in GBNF readme, align manual grammar examples & converters

### PR DESCRIPTION
**Pull request by ochafik**
- Adds a `JSON Schemas → GBNF` section to the [grammar readme](https://github.com/ggerganov/llama.cpp/blob/master/grammars/README.md)
- Updates string rule in converters using fixes from https://github.com/ggerganov/llama.cpp/pull/5888
- Updates manual JSON grammar examples w/ fixes from https://github.com/ggerganov/llama.cpp/pull/6555 (capped numeric precision) and new repetition syntax (https://github.com/ggerganov/llama.cpp/pull/6640)

cc/ @HanClinto @ExtReMLapin

### CODE REVIEW

File: common/json-schema-to-grammar.cpp [hunk 1]
```diff
@@ -57,7 +57,7 @@ std::unordered_map<std::string, BuiltinRule> PRIMITIVE_RULES = {
     {"object", {"\"{\" space ( string \":\" space value (\",\" space string \":\" space value)* )? \"}\" space", {"string", "value"}}},
     {"array", {"\"[\" space ( value (\",\" space value)* )? \"]\" space", {"value"}}},
     {"uuid", {"\"\\\"\" [0-9a-fA-F]{8} \"-\" [0-9a-fA-F]{4} \"-\" [0-9a-fA-F]{4} \"-\" [0-9a-fA-F]{4} \"-\" [0-9a-fA-F]{12} \"\\\"\" space", {}}},
-    {"char",   {"[^\"\\\\] | \"\\\\\" ([\"\\\\/bfnrt] | \"u\" [0-9a-fA-F]{4})", {}}},
+    {"char",   {"[^\"\\\\\\x7F\\x00-\\x1F] | [\\\\] ([\"\\\\bfnrt] | \"u\" [0-9a-fA-F]{4})", {}}},
```
Comment (HanClinto) [inline common/json-schema-to-grammar.cpp:60]: Should these possibly be updated to use the [new `.` operator](https://github.com/ggerganov/llama.cpp/pull/6467), or is now not the time for that?

File: grammars/json.gbnf [hunk 2]
```diff
@@ -16,10 +16,10 @@ array  ::=
...
-    "\\" (["\\/bfnrt] | "u" [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F]) # escapes
+    "\\" (["\\bfnrt] | "u" [0-9a-fA-F]{4}) # escapes
   )* "\"" ws
 
-number ::= ("-"? ([0-9] | [1-9] [0-9]*)) ("." [0-9]+)? ([eE] [-+]? [0-9]+)? ws
+number ::= ("-"? ([0-9] | [1-9] [0-9]{0,15})) ("." [0-9]+)? ([eE] [-+]? [0-9] [1-9]{0,15})? ws
 
 # Optional space: by convention, applied in this grammar after literal chars when allowed
-ws ::= ([ \t\n] ws)?
+ws ::= [ \t\n]{,20}
```
Comment (HanClinto) [inline grammars/json.gbnf:25]: This feels like a good change -- I like constraining the output in this way. Could even consider limiting it to something more restrictive like `{,4}` or `{,8}`, but this is a good start.

File: common/json-schema-to-grammar.cpp (hunk 1 above)
Comment (HanClinto) [reply common/json-schema-to-grammar.cpp:60]: Oh I see, we can't do that, because we need to exclude backslashes from this list. Nevermind, carry on! :)

**Review by HanClinto — APPROVED** [review]
Looks good to me! 👍

File: grammars/json.gbnf (hunk 2 above)
Comment (ochafik) [reply grammars/json.gbnf]: I've drafted an updated space rule in https://github.com/ggerganov/llama.cpp/pull/7866. No matter what the bound is with this current syntax, model seem keen to misuse it. But given near-unlimited indent space only (and only 1 newline at a time), they're very sensible.

**Pull request merged**